public/app/plugins/datasource/prometheus/configuration/PromSettings.tsx

```diff
@@ -1,45 +1,152 @@
+ * Advantages: We don't need to maintain a list of every possible version for each release
+ *
+ * This function will return the closest version from PromFlavorVersions that is equal or lower to the version argument,
+ * unless the versions are a major release apart.
+ *
+ * @param version
+ * @param flavor
+ */
+const getVersionString = (version: string, flavor?: string): string | undefined => {
+  if (!flavor || !PromFlavorVersions[flavor]) {
+    return;
+  }
+  const flavorVersionValues = PromFlavorVersions[flavor];
+
+  // As long as it's assured we're using versions which are sorted, we could just filter out the values greater than the target version, and then check the last element in the array
+  const versionsLessThanOrEqual = flavorVersionValues
+    ?.filter((el) => !!el.value && semver.lte(el.value, version))
+    .map((el) => el.value);
+
+  const closestVersion = versionsLessThanOrEqual[versionsLessThanOrEqual.length - 1];
+
+  if (closestVersion) {
+    const differenceBetweenActualAndClosest = semver.diff(closestVersion, version);
+
+    // Only return versions if the target is close to the actual.
+    if (['patch', 'prepatch', 'prerelease', null].includes(differenceBetweenActualAndClosest)) {
+      return closestVersion;
+    }
+  }
+
+  return;
+};
+
+const unableToDeterminePrometheusVersion = (error?: Error): void => {
+  console.warn('Error fetching version from buildinfo API, must manually select version!', error);
+};
+
+/**
+ * I don't like the daisy chain of network requests, and that we have to save on behalf of the user, but currently
+ * the backend doesn't allow for the prometheus client url to be passed in from the frontend, so we currently need to save it
+ * to the database before consumption.
+ *
+ * Since the prometheus version fields are below the url field, we can expect users to populate this field before
+ * hitting save and test at the bottom of the page. For this case we need to save the current fields before calling the
+ * resource to auto-detect the version.
+ *
+ * @param options
+ * @param onOptionsChange
+ * @param onUpdate
+ */
+const setPrometheusVersion = (
+  options: DataSourceSettingsType<PromOptions>,
+  onOptionsChange: (options: DataSourceSettingsType<PromOptions>) => void,
+  onUpdate: (dataSource: DataSourceSettingsType<PromOptions>) => Promise<DataSourceSettingsType<PromOptions>>
+) => {
+  // This will save the current state of the form, as the url is needed for this API call to function
+  onUpdate(options)
+    .then((updatedOptions) => {
+      getBackendSrv()
+        .get(`/api/datasources/${updatedOptions.id}/resources/version-detect`)
+        .then((rawResponse: PromBuildInfoResponse) => {
+          if (rawResponse.data?.version && semver.valid(rawResponse.data?.version)) {
+            const versionString = getVersionString(rawResponse.data?.version, updatedOptions.jsonData.prometheusType);
```

Comment:
`rawResponse.data?.version` if the data in rawResponse is empty this will return `undefined`. But the function is expecting a defined variable. So it might break. 
A safer approach might be `rawResponse.data?.version ?? ''` or instead of an empty string, you can pass a default version. Up to you.

Reply:
Fixed, I refactored the variable names a bit too